File: scenarios/storage-python-encrypted-uploader/golden/key_manager.py
from __future__ import annotations

import secrets

from azure.core.exceptions import HttpResponseError, ResourceNotFoundError
from azure.keyvault.keys.crypto import CryptographyClient, KeyWrapAlgorithm


class KeyVaultOperationError(Exception):
    """Raised when Key Vault cannot wrap or unwrap a data encryption key."""


class KeyManager:
    def __init__(self, crypto_client: CryptographyClient) -> None:
        self._crypto_client = crypto_client

    @property
    def key_id(self) -> str:
        return self._crypto_client.key_id

    @staticmethod
    def generate_data_encryption_key() -> bytes:
        return secrets.token_bytes(32)

    def wrap_data_encryption_key(self, dek: bytes) -> bytes:
        try:
            return self._crypto_client.wrap_key(
                KeyWrapAlgorithm.rsa_oaep,
                dek,
            ).encrypted_key
        except ResourceNotFoundError as error:
            raise KeyVaultOperationError("Key Vault could not wrap the DEK") from error
        except HttpResponseError as error:
            raise KeyVaultOperationError("Key Vault could not wrap the DEK") from error

    def unwrap_data_encryption_key(self, wrapped_dek: bytes) -> bytes:
        try:
            return self._crypto_client.unwrap_key(
                KeyWrapAlgorithm.rsa_oaep,
                wrapped_dek,
            ).key
        except ResourceNotFoundError as error:
            raise KeyVaultOperationError("Key Vault could not unwrap the DEK") from error
        except HttpResponseError as error:
            raise KeyVaultOperationError("Key Vault could not unwrap the DEK") from error

File: scenarios/storage-typescript-encrypted-uploader/golden/src/keyManager.ts
import { RestError } from "@azure/core-rest-pipeline";
import { CryptographyClient, KeyClient } from "@azure/keyvault-keys";

type TokenCredential = ConstructorParameters<typeof KeyClient>[1];

export interface WrappedDataEncryptionKey {
  keyId: string;
  wrappedDek: Buffer;
}

export class KeyManager {
  private readonly keyClient: KeyClient;

  public constructor(
    vaultUrl: string,
    private readonly keyName: string,
    private readonly credential: TokenCredential,
  ) {
    this.keyClient = new KeyClient(vaultUrl, credential);
  }

  public async wrapDataEncryptionKey(
    dataEncryptionKey: Buffer,
  ): Promise<WrappedDataEncryptionKey> {
    try {
      const key = await this.keyClient.getKey(this.keyName);
      if (!key.id) throw new Error("The Key Vault key did not include an ID.");

      const cryptographyClient = new CryptographyClient(key.id, this.credential);
      const result = await cryptographyClient.wrapKey("RSA-OAEP", dataEncryptionKey);
      return { keyId: key.id, wrappedDek: Buffer.from(result.result) };
    } catch (error) {
      if (error instanceof RestError) {
        throw new Error(`Key Vault could not wrap the data key: ${error.message}`);
      }
      throw error;
    }
  }

  public async unwrapDataEncryptionKey(
    keyId: string,
    wrappedDek: Buffer,
  ): Promise<Buffer> {
    try {
      const cryptographyClient = new CryptographyClient(keyId, this.credential);
      const result = await cryptographyClient.unwrapKey("RSA-OAEP", wrappedDek);
      return Buffer.from(result.result);
    } catch (error) {
      if (error instanceof RestError) {
        throw new Error(`Key Vault could not unwrap the data key: ${error.message}`);
      }
      throw error;
    }
  }
}
